File: PyPy3/백준/Silver/17626. Four Squares/Four Squares.py
n = int(input())
dp = [0]*(n+1) # 합이 인덱스와 같게 되는 제곱수의 최소 개수 배열
k = 1

# dp 리스트에 제곱수 저장
while k**2 <= n:
    dp[k**2] = 1
    k += 1

for i in range(1, n+1):
    if dp[i] != 0:
        continue
    j = 1
    while j*j <= i:
        if dp[i] == 0:
            dp[i] = dp[j*j] + dp[i - j*j]
        else:
            dp[i] = min(dp[i], dp[j*j] + dp[i - j*j])
        j += 1

print(dp[n])

File: node.js/백준/Silver/17626. Four Squares/Four Squares.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim();
const n = parseInt(input);

const dp = new Array(n+1).fill(0);
let k = 1;

while (k*k <= n) {
  dp[k*k] = 1
  k++;
}

for (let i = 1; i < n+1; i++) {
  if (dp[i] !== 0) continue // 처음부터 제곱수에 해당하면 넘어가기
  j = 1
  while (j*j <= i) { // 제곱수가 아닌 수일 경우
    if (dp[i] === 0) {
      dp[i] = dp[j*j] + dp[i - j*j]
    } else {
      dp[i] = Math.min(dp[i], dp[j*j] + dp[i - j*j])
    }
    j++;
  }
}

console.log(dp[n]);